bundles/org.openhab.voice.porcupineks/src/main/java/org/openhab/voice/porcupineks/internal/PorcupineKSService.java

```diff
@@ -0,0 +1,283 @@
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.voice.porcupineks.internal;
+
+import static org.openhab.voice.porcupineks.internal.PorcupineKSConstants.SERVICE_CATEGORY;
+import static org.openhab.voice.porcupineks.internal.PorcupineKSConstants.SERVICE_ID;
+import static org.openhab.voice.porcupineks.internal.PorcupineKSConstants.SERVICE_NAME;
+import static org.openhab.voice.porcupineks.internal.PorcupineKSConstants.SERVICE_PID;
+
+import java.io.BufferedInputStream;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.URL;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledExecutorService;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.core.OpenHAB;
+import org.openhab.core.audio.AudioFormat;
+import org.openhab.core.audio.AudioStream;
+import org.openhab.core.common.ThreadPoolManager;
+import org.openhab.core.config.core.ConfigurableService;
+import org.openhab.core.voice.KSErrorEvent;
+import org.openhab.core.voice.KSException;
+import org.openhab.core.voice.KSListener;
+import org.openhab.core.voice.KSService;
+import org.openhab.core.voice.KSServiceHandle;
+import org.openhab.core.voice.KSpottedEvent;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.Constants;
+import org.osgi.service.component.ComponentContext;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import ai.picovoice.porcupine.Porcupine;
+import ai.picovoice.porcupine.PorcupineException;
+
+/**
+ * The {@link PorcupineKSService} is a keyword spotting implementation based on porcupine.
+ *
+ * @author Miguel Álvarez - Initial contribution
+ */
+@NonNullByDefault
+@Component(configurationPid = SERVICE_PID, property = Constants.SERVICE_PID + "=" + SERVICE_PID)
+@ConfigurableService(category = SERVICE_CATEGORY, label = SERVICE_NAME, description_uri = SERVICE_CATEGORY + ":"
+        + SERVICE_ID)
+public class PorcupineKSService implements KSService {
+    private final Logger logger = LoggerFactory.getLogger(PorcupineKSService.class);
+    private static final String PORCUPINE_FOLDER = Path.of(OpenHAB.getUserDataFolder(), ".porcupine").toString();
+    private static final String EXTRACTION_FOLDER = Path.of(OpenHAB.getUserDataFolder(), ".porcupine", "extracted")
+            .toString();
+    private final ScheduledExecutorService executor = ThreadPoolManager.getScheduledPool("audio");
```

Comment:
I did the same on both PRs in order to share the same one, is it better that I change it to "porcupine"?

Reply:
In case you need to create a new thread pool, you have to respect a certain naming convention. I will provide more information later. 

Sharing the same thread pool between several bindings, I am not sure this is a good idea, except of course if the pool is a common pool from the core framework. We should check if a thread pool already exists for audio/voice.

Reply:
Using the common thread pool would not be a good idea as it will consume one permanent thread for each started dialog.

Our thread naming convention was for bindings, not for services. It is `OH-binding-<thing UID>[-<differentiator>]`. So I suggesting the following name for your thread pool: `OH-voice-porcupineks`.

Reply:
I have reviewed other voice bindings and none of them seems to use an executor. What should I do with this one?